File: 000_MaxSumSubarrays/test003.c
/**
 * @file test003.c
 * @author M. Z.
 * @brief Maximum value of all subarrays
 * @version 0.1
 * @date 2025-02-13
 *
 * @copyright Copyright (c) 2025
 *
 */

#include <limits.h>
#include <math.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>

int MaximumSumSubarray(int *arr, int sizeArr);

int main(void) {
  int arr[] = {3, -2, 5, -1, 4};
  printf("Size of array is %d, and maximum vaue of all subarrays is %d\n",
         sizeof(arr) / sizeof(arr[0]),
         MaximumSumSubarray(arr, sizeof(arr) / sizeof(arr[0])));
}

int MaximumSumSubarray(int *arr, int sizeArr) {
  int ret;

  if (sizeArr == 1) {
    return arr[0];
  }

  int i, m = sizeArr / 2;
  int leftMaxSumSubarr = MaximumSumSubarray(arr, m);
  int rightMaxSumSubarr = MaximumSumSubarray(arr + m, sizeArr - m);
  int leftSum = INT_MIN, rightSum = INT_MIN, sum = 0;

  for (i = m; i < sizeArr; i++) {
    sum += arr[i];
    rightSum = sum > rightSum ? sum : rightSum;
  }

  sum = 0;
  for (i = m - 1; i >= 0; i--) {
    sum += arr[i];
    leftSum = sum > leftSum ? sum : leftSum;
  }

  ret = leftMaxSumSubarr > rightMaxSumSubarr ? leftMaxSumSubarr
                                             : rightMaxSumSubarr;
  ret = ret > (leftSum + rightSum) ? ret : (leftSum + rightSum);

  return ret;
}

File: 000_MaxSumSubarrays/test003.py
'''
 * @file test003.py
 * @author M. Z.
 * @brief Maximum value of all subarrays
 * @version 0.1
 * @date 2025-02-14
 *
 * @copyright Copyright (c) 2025
'''

def MaximumSumSubarray(arr, sizeArr):
    if sizeArr == 1:
        return arr[0]
    
    m = sizeArr//2
    leftMaxSumSubarr = MaximumSumSubarray(arr, m)
    rightMaxSumSubarr = MaximumSumSubarray(arr[m:], sizeArr - m)

    sum = 0
    leftSum = rightSum = -float('inf')
    for i in range(m, sizeArr):
        sum += arr[i]
        rightSum = sum if (sum > rightSum) else rightSum

    sum = 0
    for i in range(m-1, -1, -1):
        sum += arr[i]
        leftSum = sum if (sum > leftSum) else leftSum
    
    ret = leftMaxSumSubarr if (leftMaxSumSubarr > rightMaxSumSubarr) else rightMaxSumSubarr

    ret = ret if (ret > (leftSum + rightSum)) else (leftSum + rightSum)
    for i in range(0, sizeArr):
        ret = ret if (ret > arr[i]) else arr[i]
    return ret

arr =[3,-2,5,-1,4]

print("Size of array is", len(arr), "and maximum vaue of all subarrays is", MaximumSumSubarray(arr,len(arr)))

